src/lib/operators/jit_operator/jit_operations.hpp

```diff
@@ -0,0 +1,195 @@
+  return !std::regex_match(a, regex);
+};
+
+// The InvalidTypeCatcher acts as a fallback inplementation, if template specialization
+// fails for a type combination.
+template <typename F, typename R>
+struct InvalidTypeCatcher : F {
+  explicit InvalidTypeCatcher(F f) : F(std::move(f)) {}
+
+  using F::operator();
+
+  template <typename... Ts>
+  R operator()(const Ts...) const {
+    Fail("invalid combination of types for operation");
+  }
+};
+
+template <typename T>
+__attribute__((noinline)) void jit_compute(const T& op_func, const JitMaterializedValue& lhs,
+                                           const JitMaterializedValue& rhs, JitMaterializedValue& result) {
+  // Handle NULL values and return if either input is NULL.
+  result.is_null() = lhs.is_null() || rhs.is_null();
+  if (result.is_null()) {
+    return;
+  }
+
+  const auto func = [&](const auto& typed_lhs, const auto& typed_rhs, auto& result) -> decltype(
+      op_func(typed_lhs, typed_rhs), void()) {
+    using ResultType = decltype(op_func(typed_lhs, typed_rhs));
+    result.template as<ResultType>() = op_func(typed_lhs, typed_rhs);
+  };
+
+  const auto catching_func = InvalidTypeCatcher<decltype(func), void>(func);
+
+  // The type information from the lhs and rhs are combined into a single value for dispatching without nesting.
+  const auto combined_types = static_cast<uint8_t>(lhs.data_type()) << 8 | static_cast<uint8_t>(rhs.data_type());
+  switch (combined_types) {
+    BOOST_PP_SEQ_FOR_EACH_PRODUCT(JIT_COMPUTE_CASE, (JIT_DATA_TYPE_INFO)(JIT_DATA_TYPE_INFO))
+    default:
+      Fail("unreachable");
+  }
+}
+
+template <typename T>
+__attribute__((noinline)) JitDataType jit_compute_type(const T& op_func, const JitDataType lhs, const JitDataType rhs) {
+  const auto func = [&](const auto& typed_lhs, const auto& typed_rhs) -> decltype(op_func(typed_lhs, typed_rhs),
+                                                                                  JitDataType()) {
+    using ResultType = decltype(op_func(typed_lhs, typed_rhs));
+    // This templated function returns the JitDataType enum value for a given ResultType.
+    return jit_data_type<ResultType>();
+  };
+
+  const auto catching_func = InvalidTypeCatcher<decltype(func), JitDataType>(func);
+
+  // The type information from the lhs and rhs are combined into a single value for dispatching without nesting.
+  const auto combined_types = static_cast<uint8_t>(lhs) << 8 | static_cast<uint8_t>(rhs);
+  switch (combined_types) {
+    BOOST_PP_SEQ_FOR_EACH_PRODUCT(JIT_COMPUTE_TYPE_CASE, (JIT_DATA_TYPE_INFO)(JIT_DATA_TYPE_INFO))
+    default:
+      Fail("unreachable");
+  }
+}
+
+__attribute__((noinline)) void jit_not(const JitMaterializedValue& lhs, JitMaterializedValue& result) {
+  DebugAssert(lhs.data_type() == JitDataType::Bool && result.data_type() == JitDataType::Bool,
+              "invalid type for operation");
+  result.is_null() = lhs.is_null();
+  result.as<uint8_t>() = !lhs.as<uint8_t>();
```

Comment:
Why do you cast the result into `uint8_t`, not `bool`?

Reply:
Since we are no longer returning assignable references (see the `as` --> `get`/`set` comment above), we can finally use `bool` as the type for boolean values.